### PR TITLE
Print types when not inferrable from context

### PR DESCRIPTION
**Pull request by abadams**
It's currently hard to parse Halide Exprs in isolation, because the types of Vars, Loads, and similar depend on the context, which is not always present. This PR tracks which names have a type that can be inferred from the context, and prints the type too when they're not (and when they're not just Int(32) - having a default case cuts down on clutter). With this change a hypothetical parser would just need to track a Scope from strings to types.

Also as a drive-by I added prettier printing of chains of lets around store nodes. They can get large right now. 

This PR is related to #4233 

Here's a sample of the new IR printing excerpted from local laplacian. The most obvious change in the newlines, but note also that `beta` is now `(float32)beta`. All of the other types can be inferred from the context.

```
...
                 produce f125 {
                  consume f126 {
                   let t5569 = (((output.s0.v1.v1 + t5504)*t5105) + t5503)
                   let t5568 = (output.s0.v1.v1 + t5506)
                   for (f125.s0.v0.v0, 0, t5163) {
                    let f125.s0.v0.v28.base.s = min((f125.s0.v0.v0*8), t5131)
                    f125[ramp(f125.s0.v0.v28.base.s, 1, 8)] = 
                      let t4086 = (((f125.s0.v0.v28.base.s + output.min.0)/2) + (((f125.s0.v0.v28.base.s + output.min.0) % 2)*2)) in
                      let t4090 = ((t5568/2) + ((t5568 % 2)*2)) in
                      let t4093 = f2[ramp((f125.s0.v0.v28.base.s + t5569), 2, 4)] in
                      let t4094 = (t4093*x4(t5130)) in
                      let t4095.s = min(int32x4(t4094), x4(t5065)) in
                      let t4096 = ((max(min(int32x4((t4094*x4(256.000000f))), x4(t5106)), x4(0)) - x4(f0.v0.min_realized.s)) - (max(t4095.s, x4(0))*x4(256))) in
                      let t4097 = float32x4(max(t4095.s, x4(0))) in
                      let t4098 = (t4097*x4(t5107)) in
                      let t4103 = ((max(t4095.s, x4(0)) - x4(t5066))*x4((t5097 + 1))) in
                      let t4104 = (((f125.s0.v0.v28.base.s + output.min.0)/2) + ((((t5568/2) + t5096)*f4.stride.1) - f4.v0.min_realized)) in
                      let t4107 = (((f125.s0.v0.v28.base.s + output.min.0)/2) + (((t4090 + t5096)*f4.stride.1) - f4.v0.min_realized)) in
                      let t4108 = (float32x4((max(t4095.s, x4(0)) + x4(1)))*x4(t5107)) in
                      let t4112 = (((f125.s0.v0.v28.base.s + t5134)/2) + (((f125.s0.v0.v28.base.s + t5134) % 2)*2)) in
                      let t4115 = f2[ramp(((f125.s0.v0.v28.base.s + t5569) + 1), 2, 4)] in
                      let t4116 = (t4115*x4(t5130)) in
                      let t4117.s = min(int32x4(t4116), x4(t5065)) in
                      let t4118 = ((max(min(int32x4((t4116*x4(256.000000f))), x4(t5106)), x4(0)) - x4(f0.v0.min_realized.s)) - (max(t4117.s, x4(0))*x4(256))) in
                      let t4119 = float32x4(max(t4117.s, x4(0))) in
                      let t4120 = (t4119*x4(t5107)) in
                      let t4122 = ((max(t4117.s, x4(0)) - x4(t5066))*x4((t5097 + 1))) in
                      let t4123 = (((f125.s0.v0.v28.base.s + t5134)/2) + ((((t5568/2) + t5096)*f4.stride.1) - f4.v0.min_realized)) in
                      let t4125 = (((f125.s0.v0.v28.base.s + t5134)/2) + (((t4090 + t5096)*f4.stride.1) - f4.v0.min_realized)) in
                      let t4126 = (float32x4((max(t4117.s, x4(0)) + x4(1)))*x4(t5107)) in
                      interleave_vectors((((((f126[ramp((((((t5568/2) % 8)*t5501) + (t4086 - t5502)) + -1), 1, 4)] + (f126[ramp(((((t5568/2) % 8)*t5501) + (((f125.s0.v0.v28.base.s + output.min.0)/2) - t5502)), 1, 4)]*x4(3.000000f)))*x4(3.000000f)) + (f126[ramp((((((t4090 + 7) % 8)*t5501) + (t4086 - t5502)) + -1), 1, 4)] + (f126[ramp(((((t4090 + 7) % 8)*t5501) + (((f125.s0.v0.v28.base.s + output.min.0)/2) - t5502)), 1, 4)]*x4(3.000000f))))*x4(0.062500f)) + ((((f0[(t4096 + x4(256))] + (((t4093 - t4098)*x4((float32)beta)) + t4098)) - ((((f4[(ramp(((((((t5568/2) + t5096)*f4.stride.1) - f4.v0.min_realized) + t4086) + -1), 1, 4) + t4103)] + (f4[(ramp(t4104, 1, 4) + t4103)]*x4(3.000000f)))*x4(3.000000f)) + (f4[(ramp(((((((t4090 + t5096)*f4.stride.1) - f4.v0.min_realized) + t4086) - f4.stride.1) + -1), 1, 4) + t4103)] + (f4[(ramp((t4107 - f4.stride.1), 1, 4) + t4103)]*x4(3.000000f))))*x4(0.062500f)))*((t4097 - t4094) + x4(1.000000f))) + ((t4094 - t4097)*((f0[t4096] + (((t4093 - t4108)*x4((float32)beta)) + t4108)) - ((((f4[(ramp(((((((t5568/2) + t5096)*f4.stride.1) - f4.v0.min_realized) + t4086) + t5097), 1, 4) + t4103)] + (f4[(ramp(((t4104 + t5097) + 1), 1, 4) + t4103)]*x4(3.000000f)))*x4(3.000000f)) + (f4[(ramp(((t5097 - f4.stride.1) + ((((t4090 + t5096)*f4.stride.1) - f4.v0.min_realized) + t4086)), 1, 4) + t4103)] + (f4[(ramp((((t5097 - f4.stride.1) + t4107) + 1), 1, 4) + t4103)]*x4(3.000000f))))*x4(0.062500f)))))), (((((f126[ramp((((((t5568/2) % 8)*t5501) + (t4112 - t5502)) + -1), 1, 4)] + (f126[ramp(((((t5568/2) % 8)*t5501) + (((f125.s0.v0.v28.base.s + t5134)/2) - t5502)), 1, 4)]*x4(3.000000f)))*x4(3.000000f)) + (f126[ramp((((((t4090 + 7) % 8)*t5501) + (t4112 - t5502)) + -1), 1, 4)] + (f126[ramp(((((t4090 + 7) % 8)*t5501) + (((f125.s0.v0.v28.base.s + t5134)/2) - t5502)), 1, 4)]*x4(3.000000f))))*x4(0.062500f)) + ((((f0[(t4118 + x4(256))] + (((t4115 - t4120)*x4((float32)beta)) + t4120)) - ((((f4[(ramp(((((((t5568/2) + t5096)*f4.stride.1) - f4.v0.min_realized) + t4112) + -1), 1, 4) + t4122)] + (f4[(ramp(t4123, 1, 4) + t4122)]*x4(3.000000f)))*x4(3.000000f)) + (f4[(ramp(((((((t4090 + t5096)*f4.stride.1) - f4.v0.min_realized) + t4112) - f4.stride.1) + -1), 1, 4) + t4122)] + (f4[(ramp((t4125 - f4.stride.1), 1, 4) + t4122)]*x4(3.000000f))))*x4(0.062500f)))*((t4119 - t4116) + x4(1.000000f))) + ((t4116 - t4119)*((f0[t4118] + (((t4115 - t4126)*x4((float32)beta)) + t4126)) - ((((f4[(ramp(((((((t5568/2) + t5096)*f4.stride.1) - f4.v0.min_realized) + t4112) + t5097), 1, 4) + t4122)] + (f4[(ramp(((t4123 + t5097) + 1), 1, 4) + t4122)]*x4(3.000000f)))*x4(3.000000f)) + (f4[(ramp(((t5097 - f4.stride.1) + ((((t4090 + t5096)*f4.stride.1) - f4.v0.min_realized) + t4112)), 1, 4) + t4122)] + (f4[(ramp((((t5097 - f4.stride.1) + t4125) + 1), 1, 4) + t4122)]*x4(3.000000f))))*x4(0.062500f)))))))

                   }
                  }
                 }

...
```

### CODE REVIEW

**Comment by steven-johnson** [comment]
I like that we are adding the type, but man is this gonna get verbose quickly, and the number of parens we have is huuuuuuge. Instead of `(float32)beta`, how about a syntax like `beta:f32`?

**Comment by zvookin** [comment]
So re: parsable IR, the thing has to roundtrip. This seems to be reusing cast syntax so the output would be ambiguous between whether there's an actual Cast node or not.

**Comment by abadams** [comment]
We currently use (type)val for typed constants, so I stuck with that. The issue with beta:f32 is that ':' is an allowed character in names. We could add a space but then things like `beta : f32 * 8`  are a little confusing without also adding parens, so then you have `(beta : f32) * 8`, which seems worse.

For printed exprs, this actually adds very few new types. Nearly all things are either int32 or inferrable from the context.

Cast nodes are type(val), not (type)val.

**Comment by zvookin** [comment]
It is not clear we want to use the same thing for IR printed for debuging, the written language form of Halide IR, and serialization/deserialization. The latter two could potentially be the same.

**Comment by abadams** [comment]
That would have been wise.

**Comment by steven-johnson** [comment]
> That would have been wise.

Do we have limits on what can be on our names? If not, can we declare some (reasonable) ones? (I've never seen names outside the constraints I mentioned.)

**Comment by abadams** [comment]
camera pipe uses colons. We could probably restrict the legal name space to C identifiers + $ without pissing off users too much.

**Comment by abadams** [comment]
Failure unrelated.
